### PR TITLE
104 separate integration test

### PR DESCRIPTION
**Pull request by andrii-kovalenko-ct**
Addressing issue #104 (partially):
Re-factor unit/integration tests run. Many files changed because most of them just renamed (moved to another directory).

Real changes:

  - gradle [`test configuration`](https://github.com/commercetools/commercetools-paypal-plus-integration/pull/126/files#diff-a0f6a4043ff0650fd53be30b3402584a) and respective [`intellij-idea`](https://github.com/commercetools/commercetools-paypal-plus-integration/pull/126/files#diff-a0f6a4043ff0650fd53be30b3402584a) module config
  - added [unit](https://github.com/commercetools/commercetools-paypal-plus-integration/pull/126/files#diff-c0a37452195a2f10d8521631f0e46d4e) and [integration](https://github.com/commercetools/commercetools-paypal-plus-integration/pull/126/files#diff-fe4448048f7b7b08805f262a42a88bf4) tests short documentation

  - in [payment mapper tests](https://github.com/commercetools/commercetools-paypal-plus-integration/pull/126/files#diff-24573cb10fad8cf5d02cb3b6acdf7196) inject specific test classes, instead of entire context. This makes test more _unit_, e.g. independent from 

- adjust all integration tests to follow name pattern `*IT.java` (was mixed with `*IntegrationTest.java`)

- adjust Spring tests runner names to use `SpringRunner` only (was mixed with `SpringJUnit4ClassRunner`)

The rest changes are minor renaming

### CODE REVIEW

**Comment by butenkor** [comment]
And also good work on reducing number of runners!

**Comment by andrii-kovalenko-ct** [comment]
@butenkor 

>And also good work on reducing number of runners!

Are you talking about `SpringRunner`? If so - it is an alias to `SpringJUnit4ClassRunner`, it is just shorter

File: gradle/test-config.gradle [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            compileClasspath += main.output + testCommon.output
+            runtimeClasspath += main.output + testCommon.output
+        }
+        resources.srcDirs = ["$testsIntegrationDir/resources"]
+    }
+}
+
+configurations {
+    testCommonCompile.extendsFrom compile
+    testCommonRuntime.extendsFrom runtime
```
Comment (heshamMassoud) [inline gradle/test-config.gradle]: you can have these in two lines as a one liner:
````groovy
testCommonCompile.extendsFrom compile, runtime
````

File: gradle/test-config.gradle [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    testCompile.extendsFrom testCommonCompile
+    testRuntime.extendsFrom testCommonRuntime
```
Comment (heshamMassoud) [inline gradle/test-config.gradle]: you can have these in two lines as a one liner:
````groovy
testCompile.extendsFrom testCommonCompile, testCommonRuntime
````

File: gradle/test-config.gradle [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    testIntegrationCompile.extendsFrom testCommonCompile
+    testIntegrationRuntime.extendsFrom testCommonRuntime
```
Comment (heshamMassoud) [inline gradle/test-config.gradle]: you can have these in two lines as a one liner:
````groovy
testIntegrationCompile.extendsFrom testCommonCompile, testCommonRuntime
````

File: gradle/test-config.gradle [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    description "Run only the unit tests from ${sourceSets.test.java.srcDirs}. This is an alias for :test task"
+}
+
+// tasks ordering
+// Ensure integration tests are run after unit tests and are part of check/build
+compileTestCommonJava.dependsOn classes
+compileTestJava.dependsOn classes, testCommonClasses
+compileTestIntegrationJava.dependsOn classes, testCommonClasses
+testIntegration.mustRunAfter test
+check.dependsOn test, testIntegration
```
Comment (heshamMassoud) [inline gradle/test-config.gradle]: `check.dependsOn test` is not needed here. It's done implicitly as stated by the [documentation](https://docs.gradle.org/current/userguide/java_plugin.html#_lifecycle_tasks).

> Performs all verification tasks in the project. Depends on all verification tasks in the project, including test. Some plugins add additional verification tasks to the project.

File: gradle/test-config.gradle [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+build.dependsOn test, testIntegration
```
Comment (heshamMassoud) [inline gradle/test-config.gradle]: This line is also not needed. It's done implicitly as stated by the [documentation](https://docs.gradle.org/current/userguide/java_plugin.html#_lifecycle_tasks).

> Performs a full build of the project. Depends on check and assemble.

File: gradle/test-config.gradle [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+task testUnit(dependsOn: test) {
+    description "Run only the unit tests from ${sourceSets.test.java.srcDirs}. This is an alias for :test task"
+}
+
+// tasks ordering
+// Ensure integration tests are run after unit tests and are part of check/build
+compileTestCommonJava.dependsOn classes
+compileTestJava.dependsOn classes, testCommonClasses
+compileTestIntegrationJava.dependsOn classes, testCommonClasses
```
Comment (heshamMassoud) [inline gradle/test-config.gradle]: Can you please explain why we need the above three lines?

**Comment by heshamMassoud** [comment]
@andrii-kovalenko-ct can you please also update this section in the README as it seems it's quite outdated after the directory structure changes of the tests: https://github.com/commercetools/commercetools-paypal-plus-integration/tree/104-separate-integration-test#integration-tests

File: src/test/unit/java/com/commercetools/pspadapter/notification/validation/NotificationValidationFilterTest.java [hunk 7]
```diff
@@ -12,22 +12,21 @@
...
+
     @Test
     public void whenUrlIsNotification_shouldReturnMultipleReadRequestWrapper()
             throws IOException, ServletException {
         HttpServletRequest mockRequest = mock(HttpServletRequest.class);
         HttpServletResponse mockResponse = mock(HttpServletResponse.class);
         FilterChain mockChain = mock(FilterChain.class);
-        when(mockRequest.getServletPath()).thenReturn("/" + MAIN_TEST_TENANT_NAME + "/paypalplus/notification");
-        
+        when(mockRequest.getServletPath()).thenReturn("/notification-test-tenant-name/paypalplus/notification");
```
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/test/unit/java/com/commercetools/pspadapter/notification/validation/NotificationValidationFilterTest.java]: it was an integration test specific constant and not available in the unit tests any more, as same as not needed - it is not verified in the test.

File: gradle/wrapper/gradle-wrapper.properties [hunk 8]
```diff
@@ -2,4 +2,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-bin.zip
```
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply gradle/wrapper/gradle-wrapper.properties]: 897a13b

File: gradle/test-config.gradle (hunk 1 above)
Comment (andrii-kovalenko-ct) [reply gradle/test-config.gradle]: i think `testCommonRuntime` also should be configured to be dependent on `runtime`. Not sure mixing compile and runtime configurations is a good idea, i'd keep them separated

File: gradle/test-config.gradle (hunk 6 above)
Comment (andrii-kovalenko-ct) [reply gradle/test-config.gradle]: i was not sure the task get proper dependencies graph without this explicit configuration, but looks like having proper configuration extending above is enough to make proper tasks dependencies, so removed:
075e397

File: gradle/test-config.gradle (hunk 4 above)
Comment (andrii-kovalenko-ct) [reply gradle/test-config.gradle]: 2a4812a
ok, i missed this construction adds dependency, but not overrrides

File: src/test/common/README.md [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# Common test sources
+
+In this directory we keep classes which are reused for both unit and integration tests.
+These are kind of test util methods and resources, like 
+  - [`CompletionStageUtil#executeBlocking()`](/src/test/common/java/com/commercetools/testUtil/CompletionStageUtil.java)
+  to perform blocking execution,
```
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/test/common/README.md]: 7db7378

**Comment by andrii-kovalenko-ct** [comment]
@heshamMassoud 
>can you please also update this section in the README as it seems it's quite outdated after the directory structure changes of the tests: 

The documentation is temporary separated, so updated in dedicated branch:
https://github.com/commercetools/commercetools-paypal-plus-integration/commit/18e30afba50569f308d46f3bf2c6886f4658d367

File: src/test/unit/java/com/commercetools/pspadapter/notification/validation/NotificationValidationFilterTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/unit/java/com/commercetools/pspadapter/notification/validation/NotificationValidationFilterTest.java]: But I see that you are using the same value in two unit test files: `NotificationValidationFilterTest` and `NotificationValidationInterceptorTest`. Wouldn't it better to keep it at one place? Or is it not sure that the 2 usages would always have the same value.

File: gradle/test-config.gradle (hunk 1 above)
Comment (heshamMassoud) [reply gradle/test-config.gradle]: oh, my mistake. I thought they are both `testCommonCompile`. Please ignore it.

File: src/test/unit/java/com/commercetools/pspadapter/notification/validation/NotificationValidationFilterTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply src/test/unit/java/com/commercetools/pspadapter/notification/validation/NotificationValidationFilterTest.java]: these values (_notification-test-tenant-name_) could be different in these tests, as i mentioned - they are never verified, thus no use to unify the value to constant and make the tests dependent on each other.

**Comment by andrii-kovalenko-ct** [comment]
closes #104

**Pull request merged**